Source/Core/DolphinQt2/Config/FilesystemWidget.cpp

```diff
@@ -4,6 +4,290 @@
+
+    if (info.IsDirectory())
+      PopulateDirectory(partition, item, info);
+
+    item->setData(partition, ENTRY_PARTITION);
+    item->setData(QString::fromStdString(info.GetPath()), ENTRY_NAME);
+    item->setData(info.IsDirectory() ? TYPE_DIR : TYPE_FILE, ENTRY_TYPE);
+    root->appendRow(item);
+  }
+}
+
+static QString SelectFolder()
+{
+  return QFileDialog::getExistingDirectory(nullptr, QObject::tr("Choose the folder to extract to"));
+}
+
+void FilesystemWidget::ShowContextMenu(const QPoint&)
+{
+  auto* selection = m_tree_view->selectionModel();
+  if (!selection->hasSelection())
+    return;
+
+  auto* item = m_tree_model->itemFromIndex(selection->selectedIndexes()[0]);
+
+  QMenu* menu = new QMenu(this);
+
+  int partition_id = item->data(ENTRY_PARTITION).toInt();
+  QString path = item->data(ENTRY_NAME).toString();
+
+  int type = item->data(ENTRY_TYPE).toInt();
+
+  if ((type == TYPE_DISC && m_volume->GetPartitions().size() == 0) || type == TYPE_PARTITION)
+  {
+    menu->addAction(tr("Extract System Data..."), this, [this, partition_id] {
+      auto folder = SelectFolder();
+
+      if (!folder.isEmpty())
+        ExtractSystemData(partition_id, folder);
+    });
+  }
+
+  switch (type)
+  {
+  case TYPE_DISC:
+    menu->addAction(tr("Extract Entire Disc..."), this, [this, partition_id, path] {
+      auto folder = SelectFolder();
+
+      if (folder.isEmpty())
+        return;
+
+      auto size = m_volume->GetPartitions().size();
+
+      if (size == 0)
+      {
+        ExtractPartition(-1, folder);
+      }
+      else
+      {
+        for (unsigned int i = 0; i < size; i++)
+          ExtractPartition(i, folder);
+      }
+    });
+    break;
+  case TYPE_PARTITION:
+    menu->addAction(tr("Extract Entire Partition..."), this, [this, partition_id] {
+      auto folder = SelectFolder();
+      if (!folder.isEmpty())
+        ExtractPartition(partition_id, folder);
+    });
+    menu->addSeparator();
+    menu->addAction(tr("Check Partition Integrity"), this,
+                    [this, partition_id] { CheckIntegrity(partition_id); });
+    break;
+  case TYPE_DIR:
+    menu->addAction(tr("Extract Files..."), this, [this, partition_id, path] {
+      auto folder = SelectFolder();
+
+      if (!folder.isEmpty())
+        ExtractDirectory(partition_id, path, folder);
+    });
+    break;
+  case TYPE_FILE:
+    menu->addAction(tr("Extract File..."), this, [this, partition_id, path] {
+      auto dest = QFileDialog::getSaveFileName(this, tr("Save File to"));
+
+      if (!dest.isEmpty())
+        ExtractFile(partition_id, path, dest);
+    });
+    break;
+  }
+
+  menu->exec(QCursor::pos());
+}
+
+DiscIO::Partition FilesystemWidget::GetPartitionFromID(int id)
+{
+  return id == -1 ? DiscIO::PARTITION_NONE : m_volume->GetPartitions()[id];
+}
+
+void FilesystemWidget::ExtractPartition(int partition_id, const QString& out)
+{
+  ExtractDirectory(partition_id, QStringLiteral(""), out);
+}
```

Comment:
When extracting a partition, the system data is supposed to be extracted too.

Reply:
Done.